src/liballoc/boxed.rs

```diff
@@ -63,6 +63,27 @@
 //! T` obtained from `Box::<T>::into_raw` may be deallocated using the
 //! [`Global`] allocator with `Layout::for_value(&*value)`.
 //!
+//! `Box<T>` has the same ABI as `&mut T`. In particular, when `T: Sized`,
+//! this allows using `Box<T>` in FFI:
+//!
+//! ```c
+//! /* C header */
+//! struct Foo* foo_new(void); /* Returns ownership to the caller */
+//! void foo_delete(struct Foo*); /* Takes ownership from the caller */
+//! ```
+//!
+//! ```
+//! #[repr(C)]
+//! pub struct Foo;
+//!
+//! #[no_mangle]
+//! pub extern "C" fn foo_new() -> Box<Foo> {
+//!     Box::new(Foo)
+//! }
+//!
+//! #[no_mangle]
+//! pub extern "C" fn foo_delete(_: Option<Box<Foo>>) {}
```

Comment:
Why do we use `Option<...>` here? Being nonnull is just one of so many invariants required for this pointer to be valid (being aligned, containing valid `Foo` etc.) that it's rather unnecessary to special-case it IMO. I'd simplify example to accepting just `Box<Foo>`.

Reply:
> Why do we use Option<...> here?

I think this is intended to show that `Box<T>` has a niche.

Reply:
@RReverser, this example aims to be a faithful representation of FFI code that deals with creation and deconstruction. C APIs for this, notably `free`, generally accept null pointers and do nothing. This example is consistent with such usage.

Reply:
> C APIs for this, notably free, generally accept null pointers and do nothing. 

That's true for the generic stdlib `free` that doesn't have any of the other mentioned constraints either, as it deals with raw pointers.

However, custom type-specific APIs I've either seen or wrote myself, usually expect and accept only valid pointers, both for regular methods and for destructors, hence the surprise. I guess it's fine either way.

Reply:
I guess it depends on the library. Some examples of libraries accepting null pointers in destructor-like functions (and most/all of these functions also require that the pointer points to a valid value when not null):

[GTK gtk_tree_free_path](https://developer.gnome.org/gtk3/stable/GtkTreeModel.html#gtk-tree-path-free)
[FFmpeg av_packet_free](https://ffmpeg.org/doxygen/trunk/group__lavc__packet.html#ga1066464e7cdd1f215df6940db94e5d8e)
[libgit2 git_object_free](https://github.com/libgit2/libgit2/blob/master/src/object.c#L257)

Reply:
I edited the comment as suggested regarding the behavior on null.